src/class.c

```diff
@@ -1205,6 +1205,17 @@ void ClassDeclaration::addLocalClass(ClassDeclarations *aclasses)
     aclasses->push(this);
 }
 
+/***************************************
+ * Return true if class is an InputRange.
+ * This is defined as having the members: empty, front, popFront.
+ */
+bool ClassDeclaration::isInputRange()
+{
+    return (this->search(Loc(), Id::Ffront) &&
+            this->search(Loc(), Id::FpopFront) &&
+            this->search(Loc(), Id::Fempty));
+}
+
```

Comment:
Together with the `foreach` handling code in https://github.com/D-Programming-Language/dmd/blob/master/src/statement.c#L2094 (which btw tries range iteration if it finds `front`, but will later fail if `empty` and `popFront()` are not defined), there are now three places to recognize an input range. Can they be unified?